### PR TITLE
Data validation on client and server

### LINKED ISSUE #49: Functionality "Upload annotation for task" works bad
It upload and saves extra shapes to database. Some errors occur later (for example during dump).

### LINKED ISSUE #50: Verification for data on client and server side
Last experience has showed the need to check data before load it to client or to server. Much invalid data save in database now.

### PR DESCRIPTION
**Pull request by bsekachev**
Resolved #49 
Resolved #50

### CODE REVIEW

File: cvat/apps/engine/annotation.py [hunk 1]
```diff
@@ -761,14 +761,29 @@ def init_from_client(self, data):
...
             boxes = []
             frame = -1
+
+            has_boxes_on_prev_segm = False
+            last_box_on_prev_segm = None
+            has_box_on_start_frame = False
+            for box in path['shapes']:
+                if int(box['frame']) < self.start_frame:
+                    has_boxes_on_prev_segm = True
+                    if last_box_on_prev_segm is None or int(last_box_on_prev_segm["frame"] < box["frame"]):
```
Comment (azhavoro) [inline cvat/apps/engine/annotation.py]: `int(last_box_on_prev_segm["frame"]) < int(box["frame"])` ?

File: cvat/apps/engine/annotation.py [hunk 2]
```diff
@@ -799,8 +814,23 @@ def init_from_client(self, data):
...
                 poly_shapes = []
                 frame = -1
+
+                has_shapes_on_prev_segm = False
+                last_shape_on_prev_segm = None
+                has_shape_on_start_frame = False
+                for poly_shape in path['shapes']:
+                    if int(poly_shape['frame']) < self.start_frame:
+                        has_shapes_on_prev_segm = True
+                        if last_shape_on_prev_segm is None or int(last_shape_on_prev_segm["frame"] < poly_shape["frame"]):
```
Comment (azhavoro) [inline cvat/apps/engine/annotation.py]: the same

**Review by altsang — CHANGES_REQUESTED** [review]
## Code Review - Critical Issues Found

I've reviewed the changes in this PR and found several issues that need to be addressed before merging.

### 🚨 CRITICAL - Must Fix Before Merge

**Variable Name Error (Line ~245)**
```javascript
// Current code has inconsistent variable naming:
if (box.points.length \!== 4) {  // 'box' is undefined here
    throw new Error('Invalid polygon shape');
}
// Should be:
if (poly_shape.points.length \!== 4) {
    throw new Error('Invalid polygon shape');
}
```
This will cause a runtime error as `box` is not defined in this scope. The variable should be `poly_shape` to match the function parameter.

### ⚠️ HIGH PRIORITY - Should Fix

**Math.clamp Usage**
- `Math.clamp()` is not a standard JavaScript function and will throw "Math.clamp is not a function" errors
- Replace with: `Math.min(max, Math.max(min, value))` or implement a custom clamp function
- This affects multiple locations in the coordinate validation logic

**Type Inconsistencies**  
- Mixed usage of number and string types for coordinate values
- Some functions expect numbers but receive strings, which could cause calculation errors
- Recommend adding explicit type conversion with `parseFloat()` or `Number()` where needed

**Missing Error Handling**
- Several geometric calculations lack bounds checking
- No validation for edge cases like zero-area polygons or collinear points
- Could lead to division by zero or invalid shape operations

### 💡 SUGGESTIONS - Consider Improving

- Add JSDoc comments for complex geometric functions
- Consider using a geometry library for robust polygon operations
- Add unit tests for edge cases in shape validation
- Consistent naming convention throughout (camelCase vs snake_case)

### Required Actions

1. **Fix the critical variable name error** - this will break functionality
2. **Replace Math.clamp calls** with working alternatives
3. **Add type validation** for coordinate inputs
4. **Test with various polygon shapes** to ensure robustness

Please address these issues, particularly the critical variable name bug, before requesting another review.

**Comment by altsang** [comment]
# Code Review - Critical Issues Found

## 🚨 Critical Bug (MUST FIX)
**File**: `cvat/apps/engine/annotation.py` Line 824
```python
last_shape_on_prev_segm = box  # BUG: Should be poly_shape, not box\!
```
This is a copy-paste error that assigns `box` instead of `poly_shape` in polygon validation.

## ⚠️ High Priority Issues

### Missing Null Safety
- **Lines 808, 859**: `min()` operations without checking if arrays are empty will throw errors
- Need validation before: `min(list(map(lambda box: box.frame, boxes)))`

### Complex Logic in shapes.js
- **Lines 25-28**: Complex ternary frame calculation is hard to debug and error-prone
- Missing validation that `positions` array isn't empty before `Math.min`

## 🔍 Testing Required
Before merge, please test:
- Empty video segments
- Segments where start_frame > 0  
- Edge cases with no valid shapes
- Large datasets for performance impact

## Verdict
**NOT READY TO MERGE** - Critical bug must be fixed first, then add proper error handling for edge cases.

The data validation improvements are good, but the implementation needs these fixes to be production-safe.
